cmd/lint-panics/analyzer/analyzer.go

```diff
@@ -0,0 +1,247 @@
+}
+
+func New(ctx context.Context, logger *zap.Logger) (*analysis.Analyzer, error) {
+	cfg := Config{}
+	flags, err := cfg.ParseFlags()
+	if err != nil {
+		return nil, err
+	}
+
+	logger.Info("creating analyzer", zap.String("root", cfg.RootDir))
+
+	gopls := gopls2.NewGoplsClient(ctx, logger, cfg.RootDir)
+	processor := newAnalyzer(logger, gopls, &cfg)
+
+	analyzer := &analysis.Analyzer{
+		Name:     "logpanics",
+		Doc:      fmt.Sprintf("reports missing defer call to %s", Pattern),
+		Run:      processor.Run,
+		Flags:    flags,
+		Requires: []*analysis.Analyzer{inspect.Analyzer},
+	}
+
+	return analyzer, nil
+}
+
+func newAnalyzer(logger *zap.Logger, lsp LSP, cfg *Config) *Analyzer {
+	return &Analyzer{
+		logger: logger.Named("processor"),
+		lsp:    lsp,
+		cfg:    cfg.WithAbsolutePaths(),
+	}
+}
+
+func (p *Analyzer) Run(pass *analysis.Pass) (interface{}, error) {
+	ctx, cancel := context.WithTimeout(context.Background(), 5*time.Minute)
+	defer cancel()
+
+	inspected, ok := pass.ResultOf[inspect.Analyzer].(*inspector.Inspector)
+	if !ok {
+		return nil, errors.New("analyzer is not type *inspector.Inspector")
+	}
+
+	// Check if the node is a GoStmt (which represents a 'go' statement)
+	nodeFilter := []ast.Node{
+		(*ast.GoStmt)(nil),
+	}
+
+	// Traverse the AST to find goroutines
+	inspected.Preorder(nodeFilter, func(n ast.Node) {
+		p.ProcessNode(ctx, pass, n)
+	})
+
+	return nil, nil
+}
+
+func (p *Analyzer) ProcessNode(ctx context.Context, pass *analysis.Pass, n ast.Node) {
+	goStmt, ok := n.(*ast.GoStmt)
+	if !ok {
+		panic("unexpected node type")
+	}
+
+	switch fun := goStmt.Call.Fun.(type) {
+	case *ast.FuncLit: // anonymous function
+		pos := pass.Fset.Position(fun.Pos())
+		logger := p.logger.With(
+			utils.ZapURI(pos.Filename, pos.Line),
+			zap.Int("column", pos.Column),
+		)
+
+		logger.Debug("found anonymous goroutine")
+		if err := p.checkGoroutine(fun.Body); err != nil {
+			p.logLinterError(pass, fun.Pos(), fun.Pos(), err)
+		}
+
+	case *ast.SelectorExpr: // method call
+		pos := pass.Fset.Position(fun.Sel.Pos())
+		p.logger.Info("found method call as goroutine",
+			zap.String("methodName", fun.Sel.Name),
+			utils.ZapURI(pos.Filename, pos.Line),
+			zap.Int("column", pos.Column),
+		)
+
+		defPos, err := p.checkGoroutineDefinition(ctx, pos, pass)
+		if err != nil {
+			p.logLinterError(pass, defPos, fun.Sel.Pos(), err)
+		}
+
+	case *ast.Ident: // function call
+		pos := pass.Fset.Position(fun.Pos())
+		p.logger.Info("found function call as goroutine",
+			zap.String("functionName", fun.Name),
+			utils.ZapURI(pos.Filename, pos.Line),
+			zap.Int("column", pos.Column),
+		)
+
+		defPos, err := p.checkGoroutineDefinition(ctx, pos, pass)
+		if err != nil {
+			p.logLinterError(pass, defPos, fun.Pos(), err)
+		}
+
+	default:
+		p.logger.Error("unexpected goroutine type",
+			zap.String("type", fmt.Sprintf("%T", fun)),
+		)
+	}
+}
+
+func (p *Analyzer) parseFile(path string, pass *analysis.Pass) (*ast.File, error) {
```

Comment:
Could this be somewhere else to reduce friction to implement a new linter in the future? Seems reusable, not specific of the panic linter.

Well, I guess it might be better to wait for another custom linter to decide on what/how to best split and reuse code.

Reply:
Yeah, better to wait for the need for it 👍 
Also, this one is quite small, so I'm not sure if we need to extract it.